dispatcher/dispatcher.go

```diff
@@ -0,0 +1,128 @@
+// dispatch calls from kicking off reducers until all in-flight reducers complete.
+type Dispatcher struct {
+	store    datastore.Datastore
+	reducers map[Token]dispatch.Reducer
+	lock     sync.Mutex
+}
+
+// NewDispatcher returns a new Dispatcher. This should only be called once in an application to ensure a singleton
+// dispatcher. While it is not enforced here, it is a good idea to a singleton pattern in your own code, for example:
+// 	var (
+// 		once sync.Once
+// 		instance *Dispatcher
+// 	)
+// 	var store = ...
+//
+//	func GetDispatcher() *Dispatcher {
+// 		once.Do(func() {
+// 			singleton = NewDispatcher(store)
+// 		})
+// 		return singleton
+// 	}
+func NewDispatcher(store datastore.Datastore) *Dispatcher {
+	return &Dispatcher{
+		store:    store,
+		reducers: make(map[Token]dispatch.Reducer),
+	}
+}
+
+// Register takes a reducer to be invoked with each dispatched event and returns a token for de-registration.
+func (d *Dispatcher) Register(reducer dispatch.Reducer) Token {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	lastID++
+	id := Token(lastID)
+	d.reducers[id] = reducer
+	return id
+}
+
+// Deregister removes a reducer based on its token. If the token is invalid (i.e. no associated reducer),
+// this is a no-op.
+func (d *Dispatcher) Deregister(token Token) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	if _, ok := d.reducers[token]; !ok {
+		return ErrTokenNotFound
+	}
+	delete(d.reducers, token)
+	return nil
+}
+
+// Dispatch dispatches a payload to all registered reducers. It returns a multierror object, which may contain
+// zero (nil) or more errors. Errors from reducer callbacks can be safely ignored and/or retried (and are prefixed
+// with "warning"), whereas errors due to event persistence are "critical" and will be prefixed as such.
+func (d *Dispatcher) Dispatch(event dispatch.Event) error {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+	var result *multierror.Error
+	// Key format: <timestamp>/<entity-id>/<type>
+	// @todo: This is up for debate, its a 'fake' Event struct right now anyway
+	key := datastore.NewKey(string(event.Time())).ChildString(event.EntityID()).ChildString(event.Type())
+	// Add an Event's body to the event store as the value
+	if err := d.store.Put(key, event.Body()); err != nil {
+		return ErrPersistence
+	}
+	// Fire off reducers now that event is safely persisted
+	wg := sync.WaitGroup{}
+	wg.Add(len(d.reducers))
+	errChan := make(chan error, len(d.reducers))
+	for _, reducer := range d.reducers {
+		// Launch each reducer in a separate goroutine and send back errors
+		go func(r dispatch.Reducer) {
+			defer wg.Done()
+			if err := r.Reduce(event); err != nil {
+				errChan <- multierror.Prefix(err, "warning")
+			}
+		}(reducer)
+	}
+	wg.Wait()
+	// Close and then read from error channel to put into multierror
+	close(errChan)
+	for err := range errChan {
+		result = multierror.Append(result, err)
```

Comment:
TBH, never used multierror; seems neat.
I wonder how it would really be used by the caller. Maybe just for logging, since they can't do much about it. If it was typed-error multierror maybe we could provide more info like the token of the reducer that failed. 

Maybe later when using it we can have more insight.

Reply:
Yeh, initially did it this way so that if the caller didn't know about multierror, they'd still be good (its just an error afterall)... but since we're moving this into threads, we _could_ just return the typed multierror... might as well right?

Reply:
Can be an option. But never thought about much about multierror semantics really. 
Also, not sure what consequences has a failed reducer. (The event will retry? is this fatal to the application state (since other reducers might succeed)? maybe is about defining the reducer semantics and bounded-context of state between them in docs to avoid possible pitfalls)

Reply:
Yeh this needs to be thought through a lot more I think. In theory, it should probably be retried, but I wondered if that should be left up to the caller, or if we should provide some retry facilities directly... or... I _think_ intuitively you're right about simply documenting things. Something along the lines of, if a given reducer fails, it does not affect the other reducers or event store state. It is up to the caller to catch this, and rerun if/as needed? But dunno for sure.

Reply:
We can't guarantee that a given reducer is idempotent (it probably isn't), so rerunning reducers that _didn't_ fail is bad...

Reply:
I guess it depends on what guides you tell the developer. If isn't idempotent, I agree totally. But if that's the case, this means that the reducer is in a fatal state?

Reply:
I'm going to leave this as a default error for now, but we can expose it more explicitly later if we want.